### PR TITLE
Keep track of string literal kinds

### LINKED ISSUE #8665: Discern string quote status in the AST.
IIRC the current way we represent the difference is quite messy. Since we've just landed a breaking change in `Expr`, it would seem to be a good opportunity to switch to explicitly marking single quoted string constants in the AST?

### PR DESCRIPTION
**Pull request by Simn**
Closes #8665
This change is purely structural, i.e. there's no difference with regards to format string handling or anything. I would like to do that as a separate step.

### CODE REVIEW

File: src/core/ast.ml [hunk 1]
```diff
@@ -362,7 +367,11 @@ let parse_path s =
 let s_constant = function
 	| Int s -> s
 	| Float s -> s
-	| String s -> "\"" ^ StringHelper.s_escape s ^ "\""
+	| String(s,qs) ->
+		begin match qs with
+		| SDoubleQuotes -> "\"" ^ StringHelper.s_escape s ^ "\""
+		| SSingleQuotes -> "\"" ^ StringHelper.s_escape s ^ "\""
```
Comment (RealyUniqueName) [inline src/core/ast.ml:373]: Shouldn't this print single quotes?

**Comment by RealyUniqueName** [comment]
If we merge this we _must_ decide on non-xml for 4.0.

File: src/core/ast.ml (hunk 1 above)
Comment (Simn) [reply src/core/ast.ml:373]: Yes but we need another `s_escape` for that first.

**Comment by Simn** [comment]
> If we merge this we must decide on non-xml for 4.0.

Not necessarily, we can use the same metadata-based approach we're currently using. Although it would obviously be better to sort this out once and for all.

File: std/haxe/macro/Expr.hx [hunk 2]
```diff
@@ -65,7 +70,7 @@ enum Constant {
 	/**
 		Represents a string literal.
 	**/
-	CString(s:String);
+	CString(s:String, ?kind:StringLiteralKind);
```
Comment (back2dos) [inline std/haxe/macro/Expr.hx:73]: I wonder though what is the meaning of `null` here?

Comment (Simn) [reply std/haxe/macro/Expr.hx:73]: It defaults to double quotes. But yes that should be part of the documentation.

**Pull request merged**